In [6]:
import pandas as pd
import re
import glob

## Exploring Available Books

To start with, figure out which books were not downloaded, but are present in the metadata csv!

I made a copy of the RAW texts just in case.

In [7]:
books_list = []

for name in glob.glob('../data/raw/*'):
    books_list.append(re.findall(r'PG\d*', name)[0])

In [8]:
library = pd.read_csv('../data/metadata.csv')

In [9]:
len(library) - len(books_list)

3435

There are 3435 "books" listed in the metadata that do not get downloaded. Next up, to explore why.

In [10]:
library.loc[~library['id'].isin(books_list)]['type'].value_counts()

Sound          1104
Dataset          83
Image            33
MovingImage       7
StillImage        3
Collection        1
Text              1
Name: type, dtype: int64

Starting with those that are marked as 'type' being NaN. It is possible either the flags are incorrect (I checked it with "The King James Version of the Bible"), or there is something else going on that is causing this issue. Might have to look into NaNs a little bit more.

In [11]:
library.loc[(~library['id'].isin(books_list)) & (library['type'].isna())]

id                                              title  \
606    PG10547                                        Topsy-Turvy   
703    PG10634  The Queen of Hearts, and Sing a Song for Sixpence   
841    PG10762                   Impressions of Theophrastus Such   
923    PG10836                               The Algebra of Logic   
1106      PG10                The King James Version of the Bible   
...        ...                                                ...   
70441   PG9995  Little Journey to Puerto Rico: For Intermediat...   
70442   PG9996  "'Tis Sixty Years Since": Address of Charles F...   
70443   PG9997  France and England in North America, Part III:...   
70445   PG9999                   Harriet, the Moses of Her People   
70447     PG99           Collected Articles of Frederick Douglass   

                                   author  authoryearofbirth  \
606                          Verne, Jules             1828.0   
703                   Caldecott, Randolph             1846.0   
841                         Eliot, George             1819.0   
923                       Couturat, Louis             1868.0   
1106                                  NaN                NaN   
...                                   ...                ...   
70441               George, Marian Minnie             1865.0   
70442              Adams, Charles Francis             1835.0   
70443                    Parkman, Francis             1823.0   
70445  Bradford, Sarah H. (Sarah Hopkins)             1818.0   
70447                 Douglass, Frederick             1818.0   

       authoryearofdeath language  downloads  \
606               1905.0   ['en']        126   
703               1886.0   ['en']         44   
841               1880.0   ['en']        110   
923               1914.0   ['en']         97   
1106                 NaN   ['en']       5831   
...                  ...      ...        ...   
70441                NaN   ['en']         12   
70442             1915.0   ['en']         12   
70443             1893.0   ['en']         34   
70445             1912.0   ['en']        103   
70447             1895.0   ['en']        170   

                                                subjects type  
606    {'Science fiction, French -- Translations into...  NaN  
703    {'Picture books for children', 'Nursery rhymes...  NaN  
841    {'Authors -- Fiction', 'England -- Fiction', '...  NaN  
923    {'Logic, Symbolic and mathematical', 'Algebrai...  NaN  
1106                                           {'Bible'}  NaN  
...                                                  ...  ...  
70441          {'Puerto Rico -- Description and travel'}  NaN  
70442                             {'Philosophy, Modern'}  NaN  
70443      {'Canada -- History -- To 1763 (New France)'}  NaN  
70445  {'Slaves -- United States -- Biography', 'Afri...  NaN  
70447  {'Reconstruction (U.S. history, 1865-1877)', '...  NaN  

[2215 rows x 9 columns]

For 'Sound' it is pretty straightforward. I'm only looking for boox and not for audio files.

In [12]:
library.loc[(~library['id'].isin(books_list)) & (library['type'] == 'Sound')]

id                                              title  \
151    PG10137  Mary Had a Little Lamb: Recording taken from M...   
168    PG10152           Voice Trial - Kinetophone actor audition   
169    PG10153           Voice Trial - Kinetophone Actor Audition   
170    PG10154           Voice Trial - Kinetophone Actor Audition   
171    PG10155                    The Right of the People to Rule   
...        ...                                                ...   
70159   PG9740                               Tom Tiddler's Ground   
70160   PG9741                         The Uncommercial Traveller   
70161   PG9742                       The Wreck of the Golden Mary   
70162   PG9743                          Sketches of Young Couples   
70163   PG9744                                  David Copperfield   

                                author  authoryearofbirth  authoryearofdeath  \
151    Edison, Thomas A. (Thomas Alva)             1847.0             1931.0   
168                          Lett, Bob                NaN                NaN   
169                      Lenord, Frank                NaN                NaN   
170             Schultz, Siegfried Von                NaN                NaN   
171                Roosevelt, Theodore             1858.0             1919.0   
...                                ...                ...                ...   
70159                 Dickens, Charles             1812.0             1870.0   
70160                 Dickens, Charles             1812.0             1870.0   
70161                 Dickens, Charles             1812.0             1870.0   
70162                 Dickens, Charles             1812.0             1870.0   
70163                 Dickens, Charles             1812.0             1870.0   

      language  downloads                                           subjects  \
151     ['en']         21                       {'Nursery rhymes, American'}   
168     ['en']          4                                      {'Auditions'}   
169     ['en']          4                                      {'Auditions'}   
170     ['en']          0                                      {'Auditions'}   
171     ['en']          9  {'Progressivism (United States politics)', 'Po...   
...        ...        ...                                                ...   
70159   ['en']          6                                {'English fiction'}   
70160   ['en']          6  {'England -- Social life and customs -- 19th c...   
70161   ['en']          3  {'Sea stories', 'Shipwrecks -- Fiction', 'Gold...   
70162   ['en']          3  {'Couples -- England', 'England -- Social life...   
70163   ['en']         50  {'Bildungsromans', 'Boys -- Fiction', 'Young m...   

        type  
151    Sound  
168    Sound  
169    Sound  
170    Sound  
171    Sound  
...      ...  
70159  Sound  
70160  Sound  
70161  Sound  
70162  Sound  
70163  Sound  

[1104 rows x 9 columns]

Next up, looking into datasets. It appears the vast majority of them are genomes. There are 10 calculations of square roots and 1/pi to a million digits. And 'Moby Word Lists' is just info on gutenberg, disclaimers, etc...

In [13]:
library.loc[(~library['id'].isin(books_list)) & (library['type'] == 'Dataset')].groupby('author').count()

id  title  authoryearofbirth  authoryearofdeath  \
author                                                                  
Bonnell, Jerry T.      2      2                  0                  0   
De Forest, Norman L.   1      1                  0                  0   
Human Genome Project  72     72                  0                  0   
Kanada, Yasumasa       1      1                  1                  1   
Kerr, Stan             1      1                  0                  0   
Nemiroff, Robert J.    5      5                  0                  0   
Ward, Grady            1      1                  1                  0   

                      language  downloads  subjects  type  
author                                                     
Bonnell, Jerry T.            2          2         2     2  
De Forest, Norman L.         1          1         1     1  
Human Genome Project        72         72        72    72  
Kanada, Yasumasa             1          1         1     1  
Kerr, Stan                   1          1         1     1  
Nemiroff, Robert J.          5          5         5     5  
Ward, Grady                  1          1         1     1

Onto checking out the images! the Image contains music sheets. MovingImage contains comets video, rotating earth and 5 nuclear test videos. StillImages contain a kids story illustrated and two maps/ map images.

In [14]:
library.loc[(~library['id'].isin(books_list)) & (library['type'] == 'Image')]

id                                              title  \
1108   PG11001           String Quartet No. 05 in A major Opus 18   
1109   PG11002  String Quartet No. 11 in F minor Opus 95 "Seri...   
1944   PG11755  String Quartet No. 10 in E flat major Opus 74 ...   
2381   PG12149           String Quartet No. 03 in D major Opus 18   
2479   PG12237          String Quartet No. 16 in F major Opus 135   
2986   PG12695           String Quartet No. 04 in C minor Opus 18   
3412   PG13078     String Quartet No. 12 in E flat major Opus 127   
3413   PG13079    String Quartet No. 14 in C-sharp minor Opus 131   
3495   PG13153          String Quartet No. 15 in A minor Opus 132   
3850   PG13473      String Quartet No. 06 in B flat major Opus 18   
3980   PG13590  British Folk-Music Settings Nr. 4, "Shepherd's...   
22348  PG30156  Second Overture In Solomon: Sinfonia, The Arri...   
41610   PG4749           String Quartet No. 01 in F major Opus 18   
41622   PG4750             String Quartet No. 1 in G major, K. 80   
41633   PG4751            String Quartet No. 2 in D major, K. 155   
43832   PG4949    Brandenburg Concerto No. 2 in F Major, BWV 1047   
43844   PG4950           String Quartet No. 02 in G major Opus 18   
43855   PG4951  String Quartet No. 17 in B flat major, "Hunt",...   
46488   PG5188  String Quartet No. 57 in C Major, Op. 74, No. ...   
46499   PG5189            String Quartet in C minor, Op. 51 no. 1   
46511   PG5190    Brandenburg Concerto No. 5 in D Major, BWV 1050   
51365   PG5627  String Quartet No. 10 in E-flat Major, Opus 12...   
51443   PG5634                Große Fuge in B flat major Opus 133   
51454   PG5635            String Quartet No. 3 in G Major, K. 156   
52332   PG5714  String Quartet No. 58 in F Major, Op. 74, No. ...   
54228   PG5885     Prelude and Fugue in E-Flat Major, Op. 99 no.3   
54239   PG5886            String Quartet in F Major, Op. 74 no. 2   
56439   PG6084                  Quartet in a minor, op. 41, no. 1   
65179   PG6871           String Quartet No. 09 in C major Opus 59   
67246   PG7092           String Quartet No. 07 in F major Opus 59   
67247   PG7093           String Quartet No. 08 in E minor Opus 59   
67707   PG7507     String Quartet No. 13 in B flat major Opus 130   
69772   PG9392                  Prelude in E Minor, Op. 28, No. 4   

                         author  authoryearofbirth  authoryearofdeath  \
1108      Beethoven, Ludwig van             1770.0             1827.0   
1109      Beethoven, Ludwig van             1770.0             1827.0   
1944      Beethoven, Ludwig van             1770.0             1827.0   
2381      Beethoven, Ludwig van             1770.0             1827.0   
2479      Beethoven, Ludwig van             1770.0             1827.0   
2986      Beethoven, Ludwig van             1770.0             1827.0   
3412      Beethoven, Ludwig van             1770.0             1827.0   
3413      Beethoven, Ludwig van             1770.0             1827.0   
3495      Beethoven, Ludwig van             1770.0             1827.0   
3850      Beethoven, Ludwig van             1770.0             1827.0   
3980                        NaN                NaN                NaN   
22348   Handel, George Frideric             1685.0             1759.0   
41610     Beethoven, Ludwig van             1770.0             1827.0   
41622  Mozart, Wolfgang Amadeus             1756.0             1791.0   
41633  Mozart, Wolfgang Amadeus             1756.0             1791.0   
43832    Bach, Johann Sebastian             1685.0             1750.0   
43844     Beethoven, Ludwig van             1770.0             1827.0   
43855  Mozart, Wolfgang Amadeus             1756.0             1791.0   
46488             Haydn, Joseph             1732.0             1809.0   
46499          Brahms, Johannes             1833.0             1897.0   
46511    Bach, Johann Sebastian             1685.0             1750.0   
51365           Schubert, Franz             1797.0         

In [15]:
library.loc[(~library['id'].isin(books_list)) & (library['type'] == 'StillImage')]

id                                              title  \
1661   PG114  The Tenniel Illustrations for Carroll's Alice ...   
15515  PG239                     Radar Map of the United States   
67797  PG758                  LandSat Picture of Washington, DC   

                                                  author  authoryearofbirth  \
1661                                       Tenniel, John             1820.0   
15515                                      United States                NaN   
67797  United States. National Aeronautics and Space ...                NaN   

       authoryearofdeath language  downloads  \
1661              1914.0   ['en']        391   
15515                NaN   ['en']         27   
67797                NaN   ['en']         36   

                                             subjects        type  
1661        {"Children's stories", 'Fantasy fiction'}  StillImage  
15515                       {'United States -- Maps'}  StillImage  
67797  {'Washington (D.C.) -- Remote-sensing images'}  StillImage

And finally, Collection contains 'Project Gutenberg DVD: The July 2006 Special' and the only not downloaded text is just empty.

In [16]:
library.loc[(~library['id'].isin(books_list)) & (library['type'] == 'Collection')]

id                                         title author  \
10150  PG19159  Project Gutenberg DVD: The July 2006 Special    NaN   

       authoryearofbirth  authoryearofdeath language  downloads subjects  \
10150                NaN                NaN   ['en']         73    set()   

             type  
10150  Collection

In [17]:
library.loc[(~library['id'].isin(books_list)) & (library['type'] == 'Text')]

id title author  authoryearofbirth  authoryearofdeath language  \
69464  PG90907   NaN    NaN                NaN                NaN   ['en']   

       downloads subjects  type  
69464          1    set()  Text

In [18]:
library.loc[library['author'].str.find('Lovecraft') > -1]

id                                      title  \
22880  PG30637  Writings in the United Amateur, 1915-1922   
23804  PG31469                          The Shunned House   
44538  PG50133                         The Dunwich Horror   
64643  PG68236                    The colour out of space   
64695  PG68283                        The call of Cthulhu   
64987  PG68547                                         He   
64994  PG68553                               The festival   
67130  PG70478                             The silver key   
67139  PG70486                           The lurking fear   

                                   author  authoryearofbirth  \
22880  Lovecraft, H. P. (Howard Phillips)             1890.0   
23804  Lovecraft, H. P. (Howard Phillips)             1890.0   
44538  Lovecraft, H. P. (Howard Phillips)             1890.0   
64643  Lovecraft, H. P. (Howard Phillips)             1890.0   
64695  Lovecraft, H. P. (Howard Phillips)             1890.0   
64987  Lovecraft, H. P. (Howard Phillips)             1890.0   
64994  Lovecraft, H. P. (Howard Phillips)             1890.0   
67130                     Lovecraft, H.P.                NaN   
67139  Lovecraft, H. P. (Howard Phillips)             1890.0   

       authoryearofdeath language  downloads  \
22880             1937.0   ['en']        331   
23804             1937.0   ['en']        406   
44538             1937.0   ['en']        789   
64643             1937.0   ['en']        572   
64695             1937.0   ['en']       2045   
64987             1937.0   ['en']        187   
64994             1937.0   ['en']        247   
67130                NaN   ['en']          0   
67139             1937.0   ['en']       1169   

                                                subjects type  
22880  {'Periodicals', 'Literature -- History and cri...  NaN  
23804  {'Haunted houses -- Fiction', 'Horror tales, A...  NaN  
44538  {'American fiction -- 20th century', 'Fantasy ...  NaN  
64643  {'Extraterrestrial beings -- Fiction', 'Horror...  NaN  
64695  {'Cthulhu (Fictitious character) -- Fiction', ...  NaN  
64987  {'New York (N.Y.) -- Fiction', 'Horror tales',...  NaN  
64994  {'Horror tales', 'New England -- Fiction', 'Sh...  NaN  
67130                                              set()  NaN  
67139  {'Horror tales', 'Catskill Mountains Region (N...  NaN

## Finding out English books

In [20]:
library.loc[library['language'].str.find('en') > -1]

id                                              title  \
0      PG10000                                    The Magna Carta   
1      PG10001                                    Apocolocyntosis   
2      PG10002                        The House on the Borderland   
3      PG10003         My First Years as a Frenchwoman, 1876-1879   
4      PG10004                                       The Warriors   
...        ...                                                ...   
70443   PG9997  France and England in North America, Part III:...   
70444   PG9998                                              Poems   
70445   PG9999                   Harriet, the Moses of Her People   
70447     PG99           Collected Articles of Frederick Douglass   
70448      PG9          Abraham Lincoln's First Inaugural Address   

                                   author  authoryearofbirth  \
0                               Anonymous                NaN   
1                  Seneca, Lucius Annaeus                NaN   
2                   Hodgson, William Hope             1877.0   
3                   Waddington, Mary King             1833.0   
4           Lindsay, Anna Robertson Brown             1864.0   
...                                   ...                ...   
70443                    Parkman, Francis             1823.0   
70444                     Betham, Matilda             1776.0   
70445  Bradford, Sarah H. (Sarah Hopkins)             1818.0   
70447                 Douglass, Frederick             1818.0   
70448                    Lincoln, Abraham             1809.0   

       authoryearofdeath language  downloads  \
0                    NaN   ['en']        183   
1                   65.0   ['en']        400   
2                 1918.0   ['en']        666   
3                 1923.0   ['en']         43   
4                 1948.0   ['en']         27   
...                  ...      ...        ...   
70443             1893.0   ['en']         34   
70444             1852.0   ['en']         23   
70445             1912.0   ['en']        103   
70447             1895.0   ['en']        170   
70448             1865.0   ['en']        291   

                                                subjects type  
0      {'Constitutional history -- England -- Sources...  NaN  
1      {'Claudius, Emperor of Rome, 10 B.C.-54 A.D. -...  NaN  
2                                    {'Science fiction'}  NaN  
3      {'France -- Social life and customs', 'France ...  NaN  
4                                       {'Christianity'}  NaN  
...                                                  ...  ...  
70443      {'Canada -- History -- To 1763 (New France)'}  NaN  
70444                                         {'Poetry'}  NaN  
70445  {'Slaves -- United States -- Biography', 'Afri...  NaN  
70447  {'Reconstruction (U.S. history, 1865-1877)', '...  NaN  
70448  {'Presidents -- United States -- Inaugural add...  NaN  

[56954 rows x 9 columns]